src/producer/async.js

```diff
@@ -0,0 +1,139 @@
+    this.lingerIntervalMs = this.config.producer.lingerIntervalMs === undefined ?
+      500 : this.config.producer.lingerIntervalMs
+    this.batchSize = this.config.producer.batchSize === undefined ?
+      10 : this.config.producer.batchSize
+    this.maxRetries = this.config.producer.maxRetries === undefined ?
+      3 : this.config.producer.maxRetries
+    this.retryIntervalMs = this.config.producer.retryIntervalMs === undefined ?
+      1000 : this.config.producer.retryIntervalMs
+    this.waitIntervalMs = this.config.producer.waitIntervalMs === undefined ?
+      1000 : this.config.producer.waitIntervalMs
+    this.batch = []
+    this.currentSendEventsRef = null
+    this.waitCount = 0
+  }
+
+  send (event) {
+    this.waitCount++
+    this.batch = this.batch.concat(event)
+    if (this.batch.length >= this.batchSize) {
+      clearTimeout(this.currentSendEventsRef)
+      this.sendEvents()
+      return
+    }
+    this.currentSendEventsRef =
+      setTimeout(this.sendEvents.bind(this), this.lingerIntervalMs)
+  }
+
+  sendEvents () {
+    if (this.batch.length === 0) {
+      return
+    }
+    const events = this.batch
+    this.batch = []
+    const req = {
+      id: uuid(),
+      retry: 0,
+      events
+    }
+    this.waitCount++ // wait on this batch
+    this.waitCount -= req.events.length // stop waiting on it's individual events
+    this.__sendEvents(req, 0)
+  }
+
+  async gracefulStop () {
+    if (this.waitCount === 0) {
+      return
+    }
+    this.logger.info(`waiting on ${this.waitCount} events`)
+    await new Promise(resolve => {
+      setTimeout(async () => {
+        await this.gracefulStop()
+        resolve()
+      }, this.waitIntervalMs)
+    })
+  }
+
+  __sendEvents (req) {
+    const l = this.logger.child({
+      requestId: req.id,
+      method: this.method,
+      retryCount: req.retry,
+      size: req.events.length
+    })
+    if (req.retry > this.maxRetries) {
+      l.info('dropped events due to max retries')
+      req.events.forEach(event => this.metrics.reportDropped(event.topic))
+      this.waitCount--
+      return
+    }
+    const startTime = Date.now()
+    this.grpcClient.sendEvents(req, (err, res) => {
+      const timeElapsed = Date.now() - startTime
+      req.events.forEach(event => {
+        this.metrics.reportResponseTime(this.method, event.topic, timeElapsed, err)
+      })
+      l.child({ timeElapsed, reply: res }).debug('request processed')
+      if (err) {
+        req.events.forEach(event => {
+          this.metrics.reportFailure(this.method, event.topic, err)
+        })
+        l.child({ timeElapsed, err }).error('error processing request')
+        this.__sendEventsWithRetryDelay(req)
+        return
+      }
+      let failureIndexes = []
+      if (res && res.failureIndexes && Array.isArray(res.failureIndexes)) {
+        failureIndexes = res.failureIndexes
+      }
+      let fc = 0
+      req.events.forEach((event, i) => {
```

Comment:
The idea here is to log failed events, right? I think it would be simpler to iterate over `failureIndexes`:

```
failureIndexes.forEach(idx => {
  this.metrics.reportFailure(this.method, req.events[idx].topic, "couldn't produce event")
})
```

Reply:
it's this way because it either has to reportFailure or reportSuccess

Reply:
Right! I missed that. 
Either way, I found this approach a bit confusing. I understood it now that I've read it a few times but maybe since we won't have a huge amount of events in a single request, having an additional map would make the solution easier to read. For example:

```
// TODO: handle empty failureIndexes
const failed = failureIndexes.reduce((d, idx) => , { d[idx] = true })  // this won't be huge

req.events.forEach((event, i) => {
  if (failed[i]) {
    this.metrics.reportFailure(this.method, event.topic, "couldn't produce event")
    return
  }
  this.metrics.reportSuccess(this.method, event.topic)
}
```

What do you think?

Reply:
I'm all in to improve readability. done

Reply:
❤️